src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/PullRequestStoreImpl.java

```diff
@@ -19,63 +27,186 @@
+                .getPullRequest(fromBranch, toBranch));
     }
 
     @Override
-    public void refreshStore(String key, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
+    public void refreshStore(String projectKey, String slug, String serverId, Stream<BitbucketPullRequest> bbsPullRequests) {
         PullRequestStoreImpl.CacheKey cacheKey =
-                new PullRequestStoreImpl.CacheKey(key, slug, serverId);
-        ConcurrentLinkedQueue<BitbucketPullRequest> pullRequest = pullRequests.get(cacheKey);
-        if (pullRequest != null) {
-            pullRequest.clear();
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        AtomicLong oldestUpdate = pullRequests.getOrDefault(cacheKey, new RepositoryStore(new ConcurrentHashMap<>())).findOldestUpdate();
+
+        //bbspullrequests is ordered from newest to oldest
+        bbsPullRequests.map(pr -> {
+            if (pr.getUpdatedDate() > oldestUpdate.get()) {
+                updatePullRequest(serverId, pr);
+            }
+            return pr.getUpdatedDate();
+        }).filter(updated -> updated < oldestUpdate.get())
+                .findFirst();
+    }
+
+    @Override
+    public void removeClosedPullRequests(long date) {
+       pullRequests.values().forEach(repositoryStore -> repositoryStore.removeClosedPullRequests(date));
+    }
+
+    @Override
+    public boolean hasPRForRepository(String projectKey, String slug, String serverId) {
+        PullRequestStoreImpl.CacheKey key =
+                new PullRequestStoreImpl.CacheKey(projectKey, slug, serverId);
+        return pullRequests.getOrDefault(key, new RepositoryStore(new ConcurrentHashMap<>())).hasPR();
+    }
+
+    public static MinimalPullRequest convert(BitbucketPullRequest bbsPR) {
+        return new MinimalPullRequest(bbsPR.getId(), bbsPR.getState(), bbsPR.getFromRef().getDisplayId(),
+                bbsPR.getToRef().getDisplayId(), bbsPR.getUpdatedDate());
+    }
+
+    private static class RepositoryStore {
+
+        private static final MinimalPullRequest CLOSED_PR = new MinimalPullRequest(0, BitbucketPullState.DECLINED,
+                "", "", 0);
+        private ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests;
+
+        public RepositoryStore(
+                ConcurrentMap<String, ConcurrentMap<String, MinimalPullRequest>> pullRequests) {
+            this.pullRequests = pullRequests;
+        }
+
+        public void closePullRequest(String fromBranch, String toBranch) {
+            pullRequests.getOrDefault(fromBranch, new ConcurrentHashMap<>()).getOrDefault(toBranch, CLOSED_PR)
+                    .setState(BitbucketPullState.DELETED);
+        }
+
+        public AtomicLong findOldestUpdate() {
+            AtomicLong oldestUpdate = new AtomicLong(0);
+            pullRequests.values().stream().forEach(fromBranch -> {
```

Comment:
we can do better than this, how about 
```java
 public long findOldestUpdate() {
return pullRequests.values().stream()
.flatMap(map -> map.values().stream())
.mapToLong(MinimalPullRequest::getUpdatedDate)
.max()
.orElse(0);
}
```

 maybe you can improve it even further.